[Style] Convert the 'line-height' property to strong style types

weinig opened the pull request:
#### beeb06e098ecc29cdd2736a7eb80c8ed7055e6ce
<pre>
[Style] Convert the &apos;line-height&apos; property to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=299337">https://bugs.webkit.org/show_bug.cgi?id=299337</a>

Reviewed by Darin Adler.

Converts the &apos;line-height&apos; property to use strong style types.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/rendering/TextAutoSizing.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/style/AutosizeStatus.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/inline/StyleLineHeight.cpp: Added.
* Source/WebCore/style/values/inline/StyleLineHeight.h: Added.

Canonical link: <a href="https://commits.webkit.org/301039@main">https://commits.webkit.org/301039@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03cb94aeadfa5ee89bd0db040e344992354ce112

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124768 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131613 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76683 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126645 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45143 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53010 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94928 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62978 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36007 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111579 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75496 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/124124 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34939 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29733 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75094 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105764 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29964 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134279 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39418 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103405 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52028 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107799 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103177 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26266 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48540 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26821 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48618 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51488 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57287 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50881 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54236 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/TextAutoSizing.cpp:
```diff
@@ -151,20 +152,29 @@ auto TextAutoSizingValue::adjustTextNodeSizes() -> StillHasNodes
-        else
-            specifiedLineHeight = lineHeightLength.value();
+        int specifiedLineHeight = WTF::switchOn(lineHeightLength,
+            [&](const CSS::Keyword::Normal&) -> int {
```
Style choice, I suppose, but `-> int` isn’t needed to return a `0` as an `int`.

darinadler commented on Source/WebCore/rendering/TextAutoSizing.cpp:
```diff
@@ -151,20 +152,29 @@ auto TextAutoSizingValue::adjustTextNodeSizes() -> StillHasNodes
+            [&](const Style::LineHeight::Percentage& percentage) -> int {
+                return Style::evaluate<LayoutUnit>(percentage, LayoutUnit { fontDescription.specifiedSize() });
+            },
+            [&](const Style::LineHeight::Calc&) -> int {
```
Style choice, I suppose, but `-> int` isn’t needed to return a `0` as an `int`.

darinadler commented on Source/WebCore/style/values/inline/StyleLineHeight.cpp:
```diff
@@ -0,0 +1,150 @@
+    if (!primitiveValue)
+        return CSS::Keyword::Normal { };
+
+    return this->operator()(state, *primitiveValue, multiplier);
```
Is the `this->` needed?

weinig replied:
Turns out, no, it is not.

darinadler commented on Source/WebCore/style/values/inline/StyleLineHeight.cpp:
```diff
@@ -0,0 +1,150 @@
+    if (b.isCalculated())
+        return a.isSpecified();
+
+    return false;
```
Maybe a boolean expression here?
```suggestion
    return a.hasSameType(b) || (a.isCalculated() && b.isSpecified()) || (b.isCalculated() && a.isSpecified());
```

darinadler commented on Source/WebCore/style/values/inline/StyleLineHeight.h:
```diff
@@ -0,0 +1,69 @@
+
+    bool isNormal() const { return holdsAlternative<CSS::Keyword::Normal>(); }
+
+    unsigned valueForTextAutosizingHash() const
```
Do we really need the "text autosizing" context here? Can it just be `valueForHash`?

darinadler commented on Source/WebCore/style/values/inline/StyleLineHeight.h:
```diff
@@ -0,0 +1,69 @@
+        return switchOn(
+            [](const CSS::Keyword::Normal&) -> unsigned { return 0; },
+            [](const Fixed& fixed) -> unsigned { return static_cast<unsigned>(fixed.resolveZoom(ZoomNeeded { })); },
+            [](const Percentage& percentage) -> unsigned { return static_cast<unsigned>(percentage.value); },
```
Consider preserving all the bits since both `float` and `unsigned` are 32 bits?
```suggestion
            [](const Fixed& fixed) -> unsigned { return std::bit_cast<unsigned>(fixed.resolveZoom(ZoomNeeded { })); },
            [](const Percentage& percentage) -> unsigned { return std::bit_cast<unsigned>(percentage.value); },
```
Not sure this is `float`, or if `bit_cast` will even compile.

Do we want to do anything to cut down on collisions between similar values of the 4 different types?

Finally, why would we want to resolve the fixed value to hash it?

weinig replied:
We shouldn’t resolve them. I’ll fix that.

weinig replied:
Ideally I should hash-combine with the variant index. Will fix that too.

weinig committed:
[Style] Convert the 'line-height' property to strong style types
https://bugs.webkit.org/show_bug.cgi?id=299337

Reviewed by Darin Adler.

Converts the 'line-height' property to use strong style types.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/rendering/TextAutoSizing.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/style/AutosizeStatus.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/inline/StyleLineHeight.cpp: Added.
* Source/WebCore/style/values/inline/StyleLineHeight.h: Added.

Canonical link: https://commits.webkit.org/301039@main
